woocommerce/woocommerce-gateway-stripe
# Fix issue when adding a card in My Account page

Fixes #3259 

After we have upgraded the `STRIPE_API_VERSION` to `'2024-06-20'` in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2980, user is unable to add a new payment method (card) after clicking "Add payment method" button on the "My account > Payment methods" page when "legacy checkout experience" is enabled.

In prior versions of the Stripe API, if you didn’t specify a `payment_method_types` parameter during the creation request, Stripe would default to using the card payment method for both PaymentIntents and SetupIntents.

On August 16, 2023, Stripe updated the selection process and without the `payment_method_types` param, Setup Intent is configured to automatically accept all the payment methods enabled in the Dashboard ([details](https://docs.stripe.com/upgrades/manage-payment-methods)). As some of the enabled payment methods in Stripe dashboard, needs `return_url`, if we do not specify the `payment_method_types` param when adding a card, Stripe expects a `return_url` in the request (which causes the failure).

## Changes proposed in this Pull Request:
- Sending `payment_method_types` param with create setup intent request.

## Testing instructions
- In `develop` branch, enable the legacy checkout experience.
- As a logged in shopper, go the `My Account > Payment Methods` page.
- Click 'Add payment method' and try to add a card.
- You should encounter an error and the card will not be added.
- Go to your Stripe dashboard and check the logs. You will find a failed `setup_intent` request with the following error message-
```
A `return_url` must be specified because this Setup Intent is configured to automatically accept the payment methods enabled in the Dashboard, some of which may require a full page redirect to succeed. If you do not want to accept redirect-based payment methods, set `automatic_payment_methods[enabled]` to `true` and `automatic_payment_methods[allow_redirects]` to `never` when creating Setup Intents and Payment Intents.
```
- Checkout to `fix/3259-add-payment-legacy` branch.
- Now try to add a card again on the `My Account > Payment Methods` page.
- This time you should be able to successfully add a card.

### Additional tests
- Set store currency to Euro. 
- Enable a few payment methods including SEPA.
- As a shopper go to the `My Account > Payment Methods` page.
- Try to add a SEPA token. This should be successful.
- As a shopper, make some test purchases with card, SEPA and other payment methods. 
- Make test purchases with saved cards as well as by selecting to save a new card.
- Check for other possible regressions.
